core/store/src/trie/mem/updating.rs

```diff
@@ -0,0 +1,1281 @@
+                            UpdatedMemTrieNode::Branch { children: old_children, value: None },
+                        );
+                        // if needed, branch will be squashed at the end of the function.
+                        break;
+                    } else {
+                        let mut new_children = old_children;
+                        let child = &mut new_children[partial.at(0) as usize];
+                        let old_child_id = match child.take() {
+                            Some(node_id) => node_id,
+                            None => {
+                                // Key being deleted doesn't exist.
+                                self.place_node(
+                                    node_id,
+                                    UpdatedMemTrieNode::Branch { children: old_children, value },
+                                );
+                                return;
+                            }
+                        };
+                        let new_child_id = self.ensure_updated(old_child_id);
+                        *child = Some(OldOrUpdatedNodeId::Updated(new_child_id));
+                        self.place_node(
+                            node_id,
+                            UpdatedMemTrieNode::Branch { children: new_children, value },
+                        );
+
+                        node_id = new_child_id;
+                        partial = partial.mid(1);
+                        continue;
+                    }
+                }
+                UpdatedMemTrieNode::Extension { extension, child } => {
+                    let (common_prefix, existing_len) = {
+                        let extension_nibbles = NibbleSlice::from_encoded(&extension).0;
+                        (extension_nibbles.common_prefix(&partial), extension_nibbles.len())
+                    };
+                    if common_prefix == existing_len {
+                        let new_child_id = self.ensure_updated(child);
+                        self.place_node(
+                            node_id,
+                            UpdatedMemTrieNode::Extension {
+                                extension,
+                                child: OldOrUpdatedNodeId::Updated(new_child_id),
+                            },
+                        );
+
+                        node_id = new_child_id;
+                        partial = partial.mid(existing_len);
+                        continue;
+                    } else {
+                        // Key being deleted doesn't exist.
+                        self.place_node(
+                            node_id,
+                            UpdatedMemTrieNode::Extension { extension, child },
+                        );
+                        return;
+                    }
+                }
+            }
+        }
+
+        self.squash_nodes(path);
+    }
+
+    /// Squashes intermediate nodes that are now unnecessary, e.g. if a branch has only one child.
```

Comment:
Text nit: I'd rather say "it may be necessary to change node types for which number of children decreases. For example, branch with 1 child must become a part of extension"